include/pybind11/pybind11.h

```diff
@@ -1955,25 +1955,52 @@ inline std::pair<decltype(internals::registered_types_py)::iterator, bool> all_t
     return res;
 }
 
-template <typename Iterator, typename Sentinel, bool KeyIterator, return_value_policy Policy>
+/* There are a large number of apparently unused template arguments because
+ * each combination requires a separate py::class_ registration.
+ */
+template <typename Access, return_value_policy Policy, typename Iterator, typename Sentinel, typename ValueType, typename... Extra>
 struct iterator_state {
     Iterator it;
     Sentinel end;
     bool first_or_done;
 };
 
-PYBIND11_NAMESPACE_END(detail)
+// Note: these helpers take the iterator by non-const reference because some
+// iterators in the wild can't be dereferenced when const. C++ needs the extra parens in decltype.
+template <typename Iterator, typename ResultType = decltype((*std::declval<Iterator>()))>
+struct iterator_access {
+    using result_type = ResultType;
```

Comment:
```suggestion
template <typename Iterator, typename SFINAE = decltype((*std::declval<Iterator>()))>
struct iterator_access {
    // Note: can't use SFINAE template parameter to define result_type due to a bug in ICC
    using result_type = decltype((*std::declval<Iterator>()));
```
See if that works (will need to be applied similarly to iterator_key_access and iterator_value_access.

Reply:
I believe this fixes all three "special" compilers, but breaks every version of MSVC.

```
D:\a\pybind11\pybind11\include\pybind11/pybind11.h(1991,1): error C2440: 'return': cannot convert from 'const _Ty2' to 'const int &&' [D:\a\pybind11\pybind11\tests\pybind11_tests.vcxproj]
          with
          [
              _Ty2=int
          ]
D:\a\pybind11\pybind11\include\pybind11/pybind11.h(1991,21): message : You cannot bind an lvalue to an rvalue reference [D:\a\pybind11\pybind11\tests\pybind11_tests.vcxproj]
D:\a\pybind11\pybind11\include\pybind11/pybind11.h(1990): message : while compiling class template member function 'const int &&pybind11::detail::iterator_value_access<Iterator,const _Ty2 &&>::operator ()(Iterator &) const' [D:\a\pybind11\pybind11\tests\pybind11_tests.vcxproj]
          with
          [
              Iterator=const std::pair<int,int> *,
              _Ty2=int
          ]
```

Reply:
That is extremely weird. Is this just by adding the SFINAE template parameter (relative to what was in #3271), or is it some interaction with the additional testing you added?

I see you've tried to make a workaround with `add_lvalue_reference`. I suspect that is going to cause other problems if the iterator dereferencing doesn't produce a reference (such as an input iterator). The next thing I would try is changing `std::declval<Iterator>()` to `std::declval<Iterator &>()` (all over) since the former will have type `Iterator &&` while the latter will have type `Iterator &`. I don't see why applying **dereferencing** to an rvalue reference should give another rvalue reference though.

Reply:
I think it's the additional testing, though I could be wrong. I think after adding the tests I've not been able to get MSVC to work. SFINAE seems to be working, it's just failing inside the correct overload. This would be easier if I set this up with MSVC locally on a Windows box. With nox it might be easier to setup than in the past. `add_lvalue_reference` and adding a `&` should be very similar except for void, so I'm expecting similar results, but let's see.

If we have an MSVC solution and an everyone-else solution, that would be doable.

Reply:
> add_lvalue_reference and adding a & should be very similar except for void, so I'm expecting similar results, but let's see.

Not really - you were using add_lvalue_reference to modify the result type, whereas I'm proposing modifying the iterator type. But since I don't really understand what MSVC was doing I'm also going to wait and see. I'm trying to use compiler explorer to reproduce the issue (which was pretty handy for the ICC bug), but haven't managed it yet.

Reply:
Ah, yes, I see that now. Looks like this is working a lot better! 90% complete with no failures yet.